### PR TITLE
Update Automation Module cmdlets to support Powershell7.2

### PR DESCRIPTION
**Pull request by sushil490023**
## Checklist

- [ ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Comment by NoriZC** [comment]
Hi @sushil490023, 
- Please re-record the failed test cases.
- Do you think the `Position = 4` for RuntimeVersion is necessary? If not, please remove it because a parameter with a position larger than four is discouraged.

File: src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs [hunk 1]
```diff
@@ -43,21 +53,36 @@ protected override void AutomationProcessRecord()
             IEnumerable<Module> ret = null;
             if (!string.IsNullOrEmpty(this.Name))
             {
-                ret = new List<Module>
+                if (string.IsNullOrWhiteSpace(RuntimeVersion) || RuntimeVersion.Equals(Constants.RuntimeVersion.PowerShell51))
+                {
+                    ret = new List<Module>
                 {
                    this.AutomationClient.GetModule(this.ResourceGroupName, this.AutomationAccountName, this.Name)
                 };
```
Comment (NoriZC) [inline src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs:61]:
```suggestion
                    {
                       this.AutomationClient.GetModule(this.ResourceGroupName, this.AutomationAccountName, this.Name)
                    };
```
Suitable indents will make the code more readable

File: src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs [hunk 2]
```diff
@@ -43,21 +53,36 @@ protected override void AutomationProcessRecord()
...
+                }
+                else
+                {
+                    ret = new List<Module>
+                {
+                   this.AutomationClient.GetPowerShell72Module(this.ResourceGroupName, this.AutomationAccountName, this.Name)
+                };
```
Comment (NoriZC) [inline src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs:68]:
```suggestion
                    {
                        this.AutomationClient.GetPowerShell72Module(this.ResourceGroupName, this.AutomationAccountName, this.Name)
                    };
```

File: src/Automation/Automation/Cmdlet/NewAzureAutomationModule.cs [hunk 3]
```diff
@@ -44,14 +45,34 @@ public class NewAzureAutomationModule : AzureAutomationBaseCmdlet
         [ValidateNotNullOrEmpty]
         public Uri ContentLinkUri { get; set; }
 
+
+        /// <summary>
+        /// Gets or sets the runbook version type
+        /// </summary>
+        [Parameter(Position = 4,Mandatory = false, HelpMessage = "Runtime Environment of module ")]
```
Comment (NoriZC) [inline src/Automation/Automation/Cmdlet/NewAzureAutomationModule.cs:52]:
```suggestion
        [Parameter(Mandatory = false, HelpMessage = "Runtime Environment of module ")]
```
Please help confirm whether the `Position` is needed

File: src/Automation/Automation/Cmdlet/NewAzureAutomationModule.cs [hunk 4]
```diff
@@ -44,14 +45,34 @@ public class NewAzureAutomationModule : AzureAutomationBaseCmdlet
...
+            if(string.IsNullOrWhiteSpace(RuntimeVersion) || RuntimeVersion.Equals(Constants.RuntimeVersion.PowerShell51))
+            {
+                var createdModule = this.AutomationClient.CreateModule(this.ResourceGroupName, this.AutomationAccountName, ContentLinkUri, Name);
+                this.WriteObject(createdModule);
+            }
+            else
+            {
+                var createdModule = this.AutomationClient.CreatePowerShell72Module(this.ResourceGroupName, this.AutomationAccountName, ContentLinkUri, Name);
+                this.WriteObject(createdModule);
+            }
```
Comment (NoriZC) [inline src/Automation/Automation/Cmdlet/NewAzureAutomationModule.cs:74]:
```suggestion
            if(string.IsNullOrWhiteSpace(RuntimeVersion) || RuntimeVersion.Equals(Constants.RuntimeVersion.PowerShell51))
            {
                var createdModule = this.AutomationClient.CreateModule(this.ResourceGroupName, this.AutomationAccountName, ContentLinkUri, Name);
            }
            else
            {
                var createdModule = this.AutomationClient.CreatePowerShell72Module(this.ResourceGroupName, this.AutomationAccountName, ContentLinkUri, Name);
            }          
            this.WriteObject(createdModule);
```

File: src/Automation/Automation/Common/AutomationPSClient.cs [hunk 5]
```diff
@@ -448,6 +450,129 @@ public void DeleteModule(string resourceGroupName, string automationAccountName,
...
+            {
+                if (cloudException.Response.StatusCode == System.Net.HttpStatusCode.NotFound)
+                {
+                    throw new ResourceNotFoundException(typeof(Module),
+                        string.Format(CultureInfo.CurrentCulture, Resources.ModuleNotFound, name));
+                }
+
+                throw;
+            }
+        }
```
Comment (NoriZC) [inline src/Automation/Automation/Common/AutomationPSClient.cs:493]: I didn't see any difference between this function and L352-L370. Could you explain why a new set of PowerShell72 is needed?

File: src/Automation/Automation/Common/AutomationPSClient.cs [hunk 6]
```diff
@@ -1863,7 +1988,7 @@ public Module GetPython3Package(string resourceGroupName, string automationAccou
                 moduleName,
                 new AutomationManagement.Models.PythonPackageCreateParameters()
                 { 
-                    ContentLink = new AutomationManagement.Models.ContentLink()
+                    ContentLink =  new AutomationManagement.Models.ContentLink()
```
Comment (NoriZC) [inline src/Automation/Automation/Common/AutomationPSClient.cs:1991]:
```suggestion
                    ContentLink = new AutomationManagement.Models.ContentLink()
```

File: src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs [hunk 7]
```diff
@@ -43,21 +53,36 @@ protected override void AutomationProcessRecord()
             IEnumerable<Module> ret = null;
             if (!string.IsNullOrEmpty(this.Name))
             {
-                ret = new List<Module>
+                if (string.IsNullOrWhiteSpace(RuntimeVersion) || RuntimeVersion.Equals(Constants.RuntimeVersion.PowerShell51))
```
Comment (NoriZC) [inline src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs:56]: This conditional statement appear many times. Could you extract it as a funtion?

**Review by NoriZC — CHANGES_REQUESTED** [review]

File: src/Automation/Automation/ChangeLog.md [hunk 8]
```diff
@@ -19,6 +19,10 @@
 -->
 ## Upcoming Release
 
+
+## Version 1.9.2
+* Updated Module operation cmdlets to support Powershell72
```
[the remark this answers is not in the document]
Comment (sushil490023) [reply src/Automation/Automation/ChangeLog.md:24]: Updated Thanks

File: src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs (hunk 1 above)
Comment (sushil490023) [reply src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs:61]: Updated Thanks

File: src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs (hunk 2 above)
Comment (sushil490023) [reply src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs:68]: Updated Thanks

File: src/Automation/Automation/Cmdlet/NewAzureAutomationModule.cs (hunk 3 above)
Comment (sushil490023) [reply src/Automation/Automation/Cmdlet/NewAzureAutomationModule.cs:52]: Removed Position parameter

File: src/Automation/Automation/Cmdlet/NewAzureAutomationModule.cs (hunk 4 above)
Comment (sushil490023) [reply src/Automation/Automation/Cmdlet/NewAzureAutomationModule.cs:74]: Updated

File: src/Automation/Automation/Common/AutomationPSClient.cs (hunk 5 above)
Comment (sushil490023) [reply src/Automation/Automation/Common/AutomationPSClient.cs:493]: I have Updated the function. Thanks for pointing it

File: src/Automation/Automation/Common/AutomationPSClient.cs (hunk 6 above)
Comment (sushil490023) [reply src/Automation/Automation/Common/AutomationPSClient.cs:1991]: Updated

File: src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs (hunk 7 above)
Comment (sushil490023) [reply src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs:56]: Updated , Exported as function.

File: src/Automation/Automation/Common/AutomationPSClient.cs [hunk 9]
```diff
@@ -448,20 +450,143 @@ public void DeleteModule(string resourceGroupName, string automationAccountName,
...
+        public Module UpdatePowerShell72Module(string resourceGroupName, string automationAccountName, string name,
+            Uri contentLinkUri, string contentLinkVersion)
+        {
+            try
+            {
+                var moduleModel =
+                this.automationManagementClient.PowerShell72Module.Get(resourceGroupName, automationAccountName, name);
+                if (contentLinkUri != null)
+                {
+                    var updateModule = this.automationManagementClient.Module.CreateOrUpdate(resourceGroupName,
```
Comment (NoriZC) [inline src/Automation/Automation/Common/AutomationPSClient.cs:522]:
```suggestion
                    var updateModule = this.automationManagementClient.PowerShell72Module.CreateOrUpdate(resourceGroupName,
```

**Comment by NoriZC** [comment]
Hi @sushil490023 
- The affetced test cases are not re-recorded. Please record them ans make sure they can work 
- As I commented above, you still have `Module` which haven't been updated as `PowerShell72Module`. Currently we cannot determine you code correctness by reviewing. Could you design a set of e2e tests to cover the senario?
- The difference between function Create() and CreatePowerShell72Module() is small, which means there are lots of unnecessary overlaps. Could you refine the code to make it simpler? (*For example, can we add a optional parameter into the create() to flag PowerShell 7.2?*)

File: src/Automation/Automation/Common/AutomationPSClient.cs (hunk 9 above)
Comment (sushil490023) [reply src/Automation/Automation/Common/AutomationPSClient.cs:522]: Yes, its should be sorry i missed it. I have Updated the code. Thanks for poiting it out.

**Comment by sushil490023** [comment]
Hi @NoriZC ,

1. I have recorded the test cases impacted by swagger changes in [Azure/azure-rest-api-specs@9807153](https://github.com/Azure/azure-rest-api-specs/pull/9807153). The test cases for `UpdateManagementTests` and `UpdateDynamicGroupPrePostTests` are not recorded as there have been no changes in the SDK, apart from the API version change.

image

Recording these end-to-end (E2E) test cases requires setting up resources, which will take some time. We are actively working on adding these test cases with a subsequent pull request (PR) along with Abhimanyu Varma and Sanjeev Kumar.

2. The existing set of E2E test cases currently verifies that the module is created and updated correctly. However, the response object of the cmdlets (i.e., Module) does not have any distinct parameter to distinguish whether it's PowerShell 5 or 7. I have added an explicit assertion to check that no PowerShell 7.2 modules are created or imported in PowerShell 5.1 operations, and vice versa for PowerShell 5.1.

3. Further simplification in `AutomationPSClient.cs` would impact the readability of the code with the addition of if-else statements, as for most lines of code, "Module" needs to be replaced with "PowerShell72Module."

Thanks

**Comment by NoriZC** [comment]
Hi @sushil490023
1. Big thanks that you can understand our practice of keeping the test cases running for code quality.  For the currently unrecorded and failed cases, please 
    - confirm that the behavior is the same as before after upgrading the API version.  
    - give an ETA of recording. 

    After that you can skip them by adding `[Fact(Skip = "<the reason why this is temporarily skipped>")]`. [Here](https://github.com/Azure/azure-powershell/blob/fcbc7a0061b05f7c678a04fb6b8bd099c2914c43/src/Automation/Automation.Test/ScenarioTests/JobScheduleTests.cs#L34C9-L34C90) is an example.
2. The e2e tests look good.
3. I understand your concern. But as shown below, the two method has little differences. Except the line I flagged, other code are duplicated unnecessarily. We should reduce the duplication of code. If you think the `if-else` is unnecessarily introduced as  I suggested, we can discuss a better one. But at least the current duplication is unexpected as well. 

```diff
public Module GetModule(string resourceGroupName, string automationAccountName, string name)
        {
            try
            {
                var module =
+                    this.automationManagementClient.Module.Get(resourceGroupName, automationAccountName, name);
                return new Module(resourceGroupName, automationAccountName, module);
            }
            catch (ErrorResponseException cloudException)
            {
                if (cloudException.Response.StatusCode == System.Net.HttpStatusCode.NotFound)
                {
                    throw new ResourceNotFoundException(typeof(Module),
                        string.Format(CultureInfo.CurrentCulture, Resources.ModuleNotFound, name));
                }

                throw;
            }
        }
```

```diff
public Module GetPowerShell72Module(string resourceGroupName, string automationAccountName, string name)
        {
            try
            {
                var module =
+                   this.automationManagementClient.PowerShell72Module.Get(resourceGroupName, automationAccountName, name);
                return new Module(resourceGroupName, automationAccountName, module);
            }
            catch (ErrorResponseException cloudException)
            {
                if (cloudException.Response.StatusCode == System.Net.HttpStatusCode.NotFound)
                {
                    throw new ResourceNotFoundException(typeof(Module),
                        string.Format(CultureInfo.CurrentCulture, Resources.ModuleNotFound, name));
                }

                throw;
            }
        }
```

4. Please follow the second point I mentioned in this comment to check the `Position` in parameter. I will locate one case and raise a suggestion for you. We are not encouraging position parameter any more now. 

> Hi @sushil490023 To pass the CI pipelines:
> 
> * Please re-record the failed test cases.
> * Do you think the `Position = 4` for RuntimeVersion is necessary? If not, please remove it because a parameter with a position larger than four is discouraged.

File: src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs [hunk 10]
```diff
@@ -34,6 +34,16 @@ public class GetAzureAutomationModule : AzureAutomationBaseCmdlet
         [ValidateNotNullOrEmpty]
         public string Name { get; set; }
 
+        /// <summary>
+        /// Gets or sets the runbook version type
+        /// </summary>
+        [Parameter(Position = 4, Mandatory = false, HelpMessage = "Runtime Environment of module ")]
```
Comment (NoriZC) [inline src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs:40]:
```suggestion
        [Parameter(Mandatory = false, HelpMessage = "Runtime Environment of module ")]
```

Comment (sushil490023) [reply src/Automation/Automation/Cmdlet/GetAzureAutomationModule.cs:40]: Removed

**Comment by sushil490023** [comment]
Hi @NoriZC 

1. I can confirm that no changes in skipped test cases apart from API versions changes and ETA for recording test case is 12/5.
2. I have refactored the code now to use overlapping code.
3. Removed positional parameter from all effected cmdlets.

Thanks

**Review by NoriZC — APPROVED** [review]

**Pull request merged**